### ISSUE TITLE
PaymentIntent could not be captured because it has a status of requires_action

### ISSUE THREAD

**Issue by rezaulhreza** [cartalyst/stripe#211]
I am getting the below error
 ```This PaymentIntent could not be captured because it has a status of requires_action. Only a PaymentIntent with one of the following statuses may be captured: requires_capture. ```
This is my code to create a 3d secure payment gateaway using stripe.
```
if ($this->paymentMethod == 'pay') {
            $stripe = Stripe::make(env('STRIPE_SK'));

            try {
                $token = $stripe->tokens()->create([
                    'card' => [

                        'number' => $this->card_no,
                        'exp_month' => $this->exp_month,
                        'exp_year' => $this->exp_year,
                        'cvc' => $this->cvc,

                    ]
                ]);

                if (!isset($token['id'])) {
                    session()->flash('stripe_error', 'Token not generated');
                    $this->order_success = 0;
                }
                $customer = $stripe->customers()->create([
                    'name' => $this->fname . ' ' . $this->lname,
                    'email' => $this->email,
                    'phone' => $this->contact,
                    'address' => [
                        'line1' => $this->door_number . ' ' . $this->road_name,
                        'postal_code' => $this->post_code,
                        'state' =>  $this->address,
                        'city' => $this->city,
                        'country' => $this->country,
                    ],
                    'shipping' => [
                        'name' => $this->fname . ' ' . $this->lname,
                        // 'email' => $this->email,
                        'phone' => $this->contact,
                        'address' => [
                            'line1' => $this->door_number . ' ' . $this->road_name,
                            'postal_code' => $this->post_code,
                            'state' =>  $this->address,
                            'city' => $this->city,
                            'country' => $this->country,
                        ],
                    ],
                    'source' => $token['id']

                ]);

                $paymentIntent = $stripe->paymentIntents()->create([
                    'customer' => $customer['id'],
                    'currency' => 'GBP',
                    'amount' => session()->get('checkout')['total'],
                    'description' => 'Paid for Order No. ' . $order->id,
                    'confirm' => true,
                    'capture_method' => 'automatic',
                    'confirmation_method' => 'automatic',



                ]);
                $charge = $stripe->paymentIntents()->capture($paymentIntent['id']);

                if ($paymentIntent['status'] == 'succeeded') {
                    $this->makeTransaction($order->id, 'approved');
                    $this->resetCart();
                } else {
                    session()->flash('stripe_error', 'Payment not done');
                    $this->order_success = 0;
                }
            } catch (Exception $e) {
                # code...
                session()->flash('stripe_error', $e->getMessage());
                $this->order_success = 0;
            }
        }```

**Comment by brunogaspar** [comment]
Have you tried to confirm the payment intent before calling the `capture()` method?

**Comment by rezaulhreza** [comment]
Can you please guide me how to achieve that?

**Comment by brunogaspar** [comment]
I see your passing `confirm => true` but i'm unsure why it's giving a require actions error, so trying to confirm "manually" after the payment intent creating could help (or not):

```php
... your previous code

$stripe->paymentIntents()->confirm($paymentIntent['id']);

$charge = $stripe->paymentIntents()->capture($paymentIntent['id']);

... rest of your code
```

You can also try to change/remove some of the parameters you send like the confirm, confirmation_method, etc.. and check the status of the payment intent after those changes.

If the above doesn't help, you'll need to check your Stripe logs to determine if those give an extra insight of what might be missing or being sent incorrectly, otherwise contact Stripe directly as i'm unfortunately not able to give an exact answer to what might be missing or wrong here 😞

**Comment by rezaulhreza** [comment]
after doing the below
```
                $paymentIntent = $stripe->paymentIntents()->create([
                    'customer' => $customer['id'],
                    'currency' => 'GBP',
                    'amount' => session()->get('checkout')['total'],
                    'description' => 'Paid for Order No. ' . $order->id,
                    'confirm' => false,
                    'payment_method_types' => ['card']
                   




                ]);
                $stripe->paymentIntents()->confirm($paymentIntent['id']);
                $charge = $stripe->paymentIntents()->capture($paymentIntent['id']);````

I am getting this
```You cannot confirm this PaymentIntent because it's missing a payment method. To confirm the PaymentIntent with cus_LHf56fz8aWpfd5, specify a payment method attached to this customer along with the customer ID.```

**Comment by rezaulhreza** [comment]
ok now that part is gone,
i am getting a new error now. But the weird thing is it is updating on the stripe dashboard and mark as paid. But on the chekout page, it is throwing error
```This PaymentIntent could not be captured because it has already been captured.```

I have updated the code to below:

```paymentIntent = $stripe->paymentIntents()->create([
                    'customer' => $customer['id'],
                    'currency' => 'GBP',
                    'amount' => session()->get('checkout')['total'],
                    'description' => 'Paid for Order No. ' . $order->id,
                    'confirm' => false,
                    // 'payment_method_types' => ['card']
                    // 'capture_method' => 'automatic',
                    // 'confirmation_method' => 'manual',




                ]);
                $stripe->paymentIntents()->confirm($paymentIntent['id'], ['payment_method' => 'pm_card_visa']);
                $charge = $stripe->paymentIntents()->capture($paymentIntent['id']);
```

**Comment by brunogaspar** [comment]
Give me a few i'll try to reproduce and sort it for you

**Comment by rezaulhreza** [comment]
Thanks I will wait.

Star for that

**Comment by brunogaspar** [comment]
Try the following:

```php
if ($this->paymentMethod == 'pay') {
    $stripe = Stripe::make(env('STRIPE_SK'));

    try {
        $token = $stripe->tokens()->create([
            'card' => [
                'number' => $this->card_no,
                'exp_month' => $this->exp_month,
                'exp_year' => $this->exp_year,
                'cvc' => $this->cvc,
            ],
        ]);

        if (!isset($token['id'])) {
            session()->flash('stripe_error', 'Token not generated');
            $this->order_success = 0;
        }
        $customer = $stripe->customers()->create([
            'name' => $this->fname . ' ' . $this->lname,
            'email' => $this->email,
            'phone' => $this->contact,
            'address' => [
                'line1' => $this->door_number . ' ' . $this->road_name,
                'postal_code' => $this->post_code,
                'state' =>  $this->address,
                'city' => $this->city,
                'country' => $this->country,
            ],
            'shipping' => [
                'name' => $this->fname . ' ' . $this->lname,
                // 'email' => $this->email,
                'phone' => $this->contact,
                'address' => [
                    'line1' => $this->door_number . ' ' . $this->road_name,
                    'postal_code' => $this->post_code,
                    'state' =>  $this->address,
                    'city' => $this->city,
                    'country' => $this->country,
                ],
            ],
        ]);

        $paymentIntent = $stripe->paymentIntents()->create([
            'customer' => $customer['id'],
            'currency' => 'GBP',
            'amount' => session()->get('checkout')['total'],
            'description' => 'Paid for Order No. ' . $order->id,
            'confirm' => true,
            'capture_method' => 'manual',
            'confirmation_method' => 'automatic',
            'payment_method_data' => [
                'type' => 'card',
                'card' => [
                    'token' => $token['id'],
                ],
            ],
        ]);

        $charge = $stripe->paymentIntents()->capture($paymentIntent['id']);

        if ($paymentIntent['status'] == 'succeeded') {
            $this->makeTransaction($order->id, 'approved');
            $this->resetCart();
        } else {
            session()->flash('stripe_error', 'Payment not done');
            $this->order_success = 0;
        }
    } catch (Exception $e) {
        # code...
        session()->flash('stripe_error', $e->getMessage());
        $this->order_success = 0;
    }
}
```

Basically:
- I've removed the `'source' => $token['id']` when creating the customer
- When creating the payment intent:
    - Ensured that the `capture_method` was set to manual
    - Pass the source (token you created earlier) to the payment intent via the `payment_method_data` parameter

You might need to fine tune the code but it seems to work for me.

**Comment by brunogaspar** [comment]
If you keep the auto capture when creating the payment intent, you don't need the charge part though (because the card would be auto captured/charged), and you should be able to access the charge id from the `$paymentIntent` response.

That could remove one extra API call to Stripe :)

**Comment by rezaulhreza** [comment]
I copied your code and got the error 
```This PaymentIntent could not be captured because it has a status of requires_action. Only a PaymentIntent with one of the following statuses may be captured: requires_capture```


 ``` $stripe->paymentIntents()->confirm($paymentIntent['id'], ['payment_method' => 'pm_card_visa']); ```

the ```pm_card_visa``` gets 424242 card (this works but only with the card ending 4242(testing card).
Of course I want to get the card value from the user input.

I tried to confirm the payment intent before calling the capture() method. 
```
$paymentIntent = $stripe->paymentIntents()->create([
                    'customer' => $customer['id'],
                    'currency' => 'GBP',
                    'amount' => session()->get('checkout')['total'],
                    'description' => 'Paid for Order No. ' . $order->id,
                    'confirm' => true,
                    'capture_method' => 'manual',
                    'confirmation_method' => 'automatic',
                    // 'automatic_payment_methods' => ['enabled' => true],
                    'payment_method_types' => [

                        'card',


                    ],
                ]);

                $stripe->paymentIntents()->confirm($paymentIntent['id'], ['payment_method' => 'card']);

                $charge = $stripe->paymentIntents()->capture($paymentIntent['id']);

```

When added I am getting:
You cannot confirm this PaymentIntent because it's missing a payment method. You can either update the PaymentIntent with a payment method and then confirm it again, or confirm it again directly with a payment method.

So what we understood is,it was something with the confirm intent part. So i did this:
``` $paymentIntent = $stripe->paymentIntents()->create([
                    'customer' => $customer['id'],
                    'currency' => 'GBP',
                    'amount' => session()->get('checkout')['total'],
                    'description' => 'Paid for Order No. ' . $order->id,
                    'confirm' => true,
                    'capture_method' => 'manual',

                    // 'automatic_payment_methods' => ['enabled' => true],
                    'payment_method_data' => [
                        'type' => 'card',
                        'card' => [
                            'token' => $token['id'],
                        ],
                    ],
                ]);

                $stripe->paymentIntents()->confirm($paymentIntent['id'], ['payment_method' => 'payment_method_data']);

                $charge = $stripe->paymentIntents()->capture($paymentIntent['id']);
````


In short, Main issue is with attaching payment method. I have tried for hours but could not resolve it.

**Comment by rezaulhreza** [comment]
For more info, my blade file loos like this... This is withina form.. as the issue is related with the card details, I have provided only them.

On a quick note:: I am using laravel livewire(Just incase if there are different approaches)
<div class="w-full mx-auto rounded-lg bg-white shadow-lg p-5 text-gray-700" style="max-width: 600px">
     
        <div class="mb-3">
            <label for="card-no" class="font-bold text-sm mb-2 ml-1">Card number</label>
            <div>
                <input name="card-no" id="card_number" wire:model="card_no" class="w-full px-3 py-2 mb-1 border-2 border-gray-200 rounded-md focus:outline-none focus:border-orange-500 transition-colors" placeholder="0000 0000 0000 0000" type="text"/>
            </div>
             @error('card_no') <span class="text-red-500">{{ $message }}</span>@enderror
        </div>
     
        
        <div class="mb-3">
            <label for="exp-month" class="font-bold text-sm mb-2 ml-1">Expiry Month</label>
            <div>
                <input name="exp-month" id="exp_month" wire:model="exp_month" class="w-full px-3 py-2 mb-1 border-2 border-gray-200 rounded-md focus:outline-none focus:border-orange-500 transition-colors" placeholder="MM" type="text"/>
            </div>
             @error('exp_month') <span class="text-red-500">{{ $message }}</span>@enderror
        </div>

        
        <div class="mb-3">
            <label for="exp-year" class="font-bold text-sm mb-2 ml-1">Expiry Year</label>
            <div>
                <input name="exp-year" id="exp_year" wire:model="exp_year" class="w-full px-3 py-2 mb-1 border-2 border-gray-200 rounded-md focus:outline-none focus:border-orange-500 transition-colors" placeholder="YYYY" type="text"/>
            </div>
             @error('exp_year') <span class="text-red-500">{{ $message }}</span>@enderror
        </div>


        
        <div class="mb-3">
            <label for="cvc" class="font-bold text-sm mb-2 ml-1">CVC</label>
            <div>
                <input name="cvc" id="cvc" wire:model="cvc" class="w-full px-3 py-2 mb-1 border-2 border-gray-200 rounded-md focus:outline-none focus:border-orange-500 transition-colors" placeholder="000" type="password"/>
            </div>
             @error('cvc') <span class="text-red-500">{{ $message }}</span>@enderror
        </div>
        
    </div>
```

**Comment by rezaulhreza** [comment]
I have tried another approach to get card details:
```$token = $stripe->tokens()->create([
                    'card' => [
                        'number' => $this->card_no,
                        'exp_month' => $this->exp_month,
                        'exp_year' => $this->exp_year,
                        'cvc' => $this->cvc,
                    ],
                ]);

                if (!isset($token['id'])) {
                    session()->flash('stripe_error', 'Token not generated');
                    $this->order_success = 0;
                }
                $customer = $stripe->customers()->create([
                    'name' => $this->fname . ' ' . $this->lname,
                    'email' => $this->email,
                    'phone' => $this->contact,
                    'address' => [
                        'line1' => $this->door_number . ' ' . $this->road_name,
                        'postal_code' => $this->post_code,
                        'state' =>  $this->address,
                        'city' => $this->city,
                        'country' => $this->country,
                    ],
                    'shipping' => [
                        'name' => $this->fname . ' ' . $this->lname,
                        // 'email' => $this->email,
                        'phone' => $this->contact,
                        'address' => [
                            'line1' => $this->door_number . ' ' . $this->road_name,
                            'postal_code' => $this->post_code,
                            'state' =>  $this->address,
                            'city' => $this->city,
                            'country' => $this->country,
                        ],
                    ],
                ]);

                $paymentIntent = $stripe->paymentIntents()->create([
                    'customer' => $customer['id'],
                    'currency' => 'GBP',
                    'amount' => session()->get('checkout')['total'],
                    'description' => 'Paid for Order No. ' . $order->id,
                    'confirm' => true,
                    'capture_method' => 'manual',

                    // 'automatic_payment_methods' => ['enabled' => true],
                    'payment_method_data' => [
                        'type' => 'card',
                        'card' => [
                            'token' => $token['id'],
                        ],
                    ],
                ]);

                $stripe->paymentIntents()->confirm($paymentIntent['id'], ['payment_method' => $token]);
                                                               /*                                                             ^^^^ added this from $token = createTokens ^^^
*/
                $charge = $stripe->paymentIntents()->capture($paymentIntent['id']);
````




---------------------------------------------
==============
---------------------------------------------


However, this time I am getting:

```Invalid string: {:id=>"tok_1KbFGMFF7RMxVxWd4zb7Fryf", :object=>"token", :card=>{:id=>"card_1KbFGMFF7RMxVxWdkce903N3", :object=>"card", :brand=>"Visa", :country=>"DE", :cvc_check=>"unchecked", :exp_month=>"12", :exp_year=>"2028", :fingerprint=>"bQzWK5OsbLrcJL78", :funding=>"credit", :last4=>"3184"}, :client_ip=>"78.145.188.85", :created=>"1646792910", :livemode=>"0", :type=>"card", :used=>"0"}```

**Comment by brunogaspar** [comment]
You don't need to confirm the payment manually anymore, since you're passing `confirm = true` when creating the payment intent, therefore, the following line can be removed:

```php
$stripe->paymentIntents()->confirm($paymentIntent['id'], ['payment_method' => $token]);
```

About your invalid string error, you're passing the entire `$token` array when confirm the payment intent, but as i mentioned, you don't need to manually confirm.

This is the code i've used and it works:

```php
$token = Stripe::tokens()->create([
    'card' => [
        'number' => '4242424242424242', // $this->card_no,
        'exp_month' => '12', // $this->exp_month,
        'exp_year' => '23', // $this->exp_year,
        'cvc' => '123', // $this->cvc,
    ],
]);

$customer = Stripe::customers()->create([
    'name' => 'John Doe', // $this->fname . ' ' . $this->lname,
    'email' => 'john@doe.com', // $this->email,
    'phone' => '+123456789', // $this->contact,
    'address' => [
        'line1' => 'Line 1', // $this->door_number . ' ' . $this->road_name,
        'postal_code' => '4444-444', // $this->post_code,
        'state' => 'State', // $this->address,
        'city' => 'City', // $this->city,
        'country' => 'United States', // $this->country,
    ],
    'shipping' => [
        'name' => 'John Doe', // $this->fname . ' ' . $this->lname,
        // 'email' => $this->email,
        'phone' => '+123456789', // $this->contact,
        'address' => [
            'line1' => 'Line 1', // $this->door_number . ' ' . $this->road_name,
            'postal_code' => '4444-444', // $this->post_code,
            'state' => 'State', // $this->address,
            'city' => 'City', // $this->city,
            'country' => 'United States', // $this->country,
        ],
    ],
]);

$paymentIntent = Stripe::paymentIntents()->create([
    'customer' => $customer['id'],
    'currency' => 'GBP',
    'amount' => '10059',
    'description' => 'Paid for Order No. ',
    'confirm' => true,
    'capture_method' => 'manual',
    'confirmation_method' => 'automatic',
    'payment_method_data' => [
        'type' => 'card',
        'card' => [
            'token' => $token['id'],
        ],
    ],
]);

$charge = Stripe::paymentIntents()->capture($paymentIntent['id']);

dd($charge); // status == succeeded
```

You should also consider using Stripe.js to generate the token on Stripe end instead of passing the card details to your server, as that's not secure and you would need to be PCI compliant in the end, but of course, that's up to you.
